File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/customcontrols/LayoutConverter.java
package net.kdt.pojavlaunch.customcontrols;

import android.content.Context;

import com.google.gson.JsonSyntaxException;
import com.arata.yukarilauncher.R;

import net.kdt.pojavlaunch.LwjglGlfwKeycode;
import net.kdt.pojavlaunch.Tools;

import org.apache.commons.io.IOUtils;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.lwjgl.glfw.CallbackBridge;

import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;

public class LayoutConverter {
/**
 * 「load And Convert If Necessary」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    public static CustomControls loadAndConvertIfNecessary(Context ctx, String jsonPath) throws IOException, JsonSyntaxException {

        String jsonLayoutData = Tools.read(jsonPath);
        try {
            JSONObject layoutJobj = new JSONObject(jsonLayoutData);
            return loadAndConvertIfNecessary(ctx, layoutJobj, jsonLayoutData, jsonPath, true);
        } catch (Exception e) {
            Tools.showError(ctx, ctx.getString(R.string.controls_load_failed), e);
            return null;
        }
    }
/**
 * 「load From Assets」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    public static CustomControls loadFromAssets(Context context, String jsonName) {
        try (InputStream is = context.getAssets().open(jsonName)) {
            String string = IOUtils.toString(is, StandardCharsets.UTF_8);

            JSONObject layoutJobj = new JSONObject(string);
            return loadAndConvertIfNecessary(context, layoutJobj, string, null, true);
        } catch (Exception e) {
            Tools.showError(context, context.getString(R.string.controls_load_failed), e);
            return null;
        }
    }
/**
 * 「load From Json Object」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    public static CustomControls loadFromJsonObject(Context ctx, JSONObject layoutJobj, String jsonString, String jsonPath, boolean showError) {
        try {
            return loadAndConvertIfNecessary(ctx, layoutJobj, jsonString, jsonPath, showError);
        } catch (Exception e) {
            if (showError) Tools.showError(ctx, ctx.getString(R.string.controls_load_failed), e);
            return null;
        }
    }
/**
 * 「load And Convert If Necessary」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    private static CustomControls loadAndConvertIfNecessary(Context ctx, JSONObject layoutJobj, String jsonLayoutData, String jsonPath, boolean showError) throws Exception {
        if (!layoutJobj.has("version")) { //v1 layout
            CustomControls layout = LayoutConverter.convertV1Layout(layoutJobj);
            if (jsonPath != null) layout.save(jsonPath);
            return layout;
        } else {
            int version = layoutJobj.getInt("version");
            if (version == 2) {
                CustomControls layout = LayoutConverter.convertV2Layout(layoutJobj);
                if (jsonPath != null) layout.save(jsonPath);
                return layout;
            }
            if (version == 3 || version == 4 || version == 5) {
                return LayoutConverter.convertV3_4Layout(layoutJobj);
            }
            if (version == 6 || version == 7) {
                return convertV6_7Layout(layoutJobj);
            }
            if (version == 8) {
                return Tools.GLOBAL_GSON.fromJson(jsonLayoutData, CustomControls.class);
            }
            if (showError) {
                IOException ioException = new IOException();
                Tools.showError(ctx, ctx.getString(R.string.controls_unsupported_layout_version), ioException);
            }
            return null;
        }
    }

    /**
     * レイアウトをv6/7からv8に正規化します。ジョイスティックの高さと位置に関する問題を修正する必要があります。
     * @param oldLayoutJson 古いレイアウト
     * @return 修正されたジョイスティック高さを持つ新しいレイアウト
     */
    public static CustomControls convertV6_7Layout(JSONObject oldLayoutJson) {
        CustomControls layout = Tools.GLOBAL_GSON.fromJson(oldLayoutJson.toString(), CustomControls.class);
        for (ControlJoystickData data : layout.mJoystickDataList) {
            if (data.getHeight() > data.getWidth()) {
                // Make the size square, adjust the dynamic position related to height
                float ratio = data.getHeight() / data.getWidth();
                data.dynamicX = data.dynamicX.replace("${height}", "(" + ratio + " * ${height})");
                data.dynamicY = data.dynamicY.replace("${height}", "(" + ratio + " * ${height})") +  " + (" + (ratio-1) + " * ${height})";
                data.setHeight(data.getWidth());
            }
        }
        layout.version = 8;
        return layout;
    }

    /**
     * レイアウトをv3/4からv6に正規化します。ストローク幅はボタンサイズに依存しなくなりました。
     */
    private static CustomControls convertV3_4Layout(JSONObject oldLayoutJson) {
        CustomControls layout = Tools.GLOBAL_GSON.fromJson(oldLayoutJson.toString(), CustomControls.class);
        convertStrokeWidth(layout);
        layout.version = 6;
        return layout;
    }
/**
 * 「convert 2 Layout」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */

    private static CustomControls convertV2Layout(JSONObject oldLayoutJson) throws JSONException {
        CustomControls layout = Tools.GLOBAL_GSON.fromJson(oldLayoutJson.toString(), CustomControls.class);
        JSONArray layoutMainArray = oldLayoutJson.getJSONArray("mControlDataList");
        layout.mControlDataList = new ArrayList<>(layoutMainArray.length());
        for (int i = 0; i < layoutMainArray.length(); i++) {
            JSONObject button = layoutMainArray.getJSONObject(i);
            ControlData n_button = Tools.GLOBAL_GSON.fromJson(button.toString(), ControlData.class);
            if (!Tools.isValidString(n_button.dynamicX) && button.has("x")) {
                double buttonC = button.getDouble("x");
                double ratio = buttonC / CallbackBridge.physicalWidth;
                n_button.dynamicX = ratio + " * ${screen_width}";
            }
            if (!Tools.isValidString(n_button.dynamicY) && button.has("y")) {
                double buttonC = button.getDouble("y");
                double ratio = buttonC / CallbackBridge.physicalHeight;
                n_button.dynamicY = ratio + " * ${screen_height}";
            }
            layout.mControlDataList.add(n_button);
        }
        JSONArray layoutDrawerArray = oldLayoutJson.getJSONArray("mDrawerDataList");
        layout.mDrawerDataList = new ArrayList<>();
        for (int i = 0; i < layoutDrawerArray.length(); i++) {
            JSONObject button = layoutDrawerArray.getJSONObject(i);
            JSONObject buttonProperties = button.getJSONObject("properties");
            ControlDrawerData n_button = Tools.GLOBAL_GSON.fromJson(button.toString(), ControlDrawerData.class);
            if (!Tools.isValidString(n_button.properties.dynamicX) && buttonProperties.has("x")) {
                double buttonC = buttonProperties.getDouble("x");
                double ratio = buttonC / CallbackBridge.physicalWidth;
                n_button.properties.dynamicX = ratio + " * ${screen_width}";
            }
            if (!Tools.isValidString(n_button.properties.dynamicY) && buttonProperties.has("y")) {
                double buttonC = buttonProperties.getDouble("y");
                double ratio = buttonC / CallbackBridge.physicalHeight;
                n_button.properties.dynamicY = ratio + " * ${screen_height}";
            }
            layout.mDrawerDataList.add(n_button);
        }
        convertStrokeWidth(layout);

        layout.version = 3;
        return layout;
    }
/**
 * 「convert 1 Layout」処理を実行します。
 * このメソッドは特定の機能を提供するために実装されています。
 */
    private static CustomControls convertV1Layout(JSONObject oldLayoutJson) throws JSONException {
        CustomControls empty = new CustomControls();
        JSONArray layoutMainArray = oldLayoutJson.getJSONArray("mControlDataList");
        for (int i = 0; i < layoutMainArray.length(); i++) {
            JSONObject button = layoutMainArray.getJSONObject(i);
            ControlData n_button = new ControlData();
            int[] keycodes = new int[]{LwjglGlfwKeycode.GLFW_KEY_UNKNOWN,
                    LwjglGlfwKeycode.GLFW_KEY_UNKNOWN,
                    LwjglGlfwKeycode.GLFW_KEY_UNKNOWN,
                    LwjglGlfwKeycode.GLFW_KEY_UNKNOWN};
            n_button.dynamicX = button.getString("dynamicX");
            n_button.dynamicY = button.getString("dynamicY");
            if (!Tools.isValidString(n_button.dynamicX) && button.has("x")) {
                double buttonC = button.getDouble("x");
                double ratio = buttonC / CallbackBridge.physicalWidth;
                n_button.dynamicX = ratio + " * ${screen_width}";
            }
            if (!Tools.isValidString(n_button.dynamicY) && button.has("y")) {
                double buttonC = button.getDouble("y");
                double ratio = buttonC / CallbackBridge.physicalHeight;
                n_button.dynamicY = ratio + " * ${screen_height}";
            }
            n_button.name = button.getString("name");
            n_button.opacity = ((float) ((button.getInt("transparency") - 100) * -1)) / 100f;
            n_button.passThruEnabled = button.getBoolean("passThruEnabled");
            n_button.isToggle = button.getBoolean("isToggle");
            n_button.setHeight(button.getInt("height"));
            n_button.setWidth(button.getInt("width"));
            n_button.bgColor = 0x4d000000;
            n_button.strokeWidth = 0;
            if (button.getBoolean("isRound")) {
                n_button.cornerRadius = 35f;
            }
            int next_idx = 0;
            if (button.getBoolean("holdShift")) {
                keycodes[next_idx] = LwjglGlfwKeycode.GLFW_KEY_LEFT_SHIFT;
                next_idx++;
            }
            if (button.getBoolean("holdCtrl")) {
                keycodes[next_idx] = LwjglGlfwKeycode.GLFW_KEY_LEFT_CONTROL;
                next_idx++;
            }
            if (button.getBoolean("holdAlt")) {
                keycodes[next_idx] = LwjglGlfwKeycode.GLFW_KEY_LEFT_ALT;
                next_idx++;
            }
            keycodes[next_idx] = button.getInt("keycode");
            n_button.keycodes = keycodes;
            empty.mControlDataList.add(n_button);
        }
        empty.scaledAt = (float) oldLayoutJson.getDouble("scaledAt");
        empty.version = 3;
        return empty;
    }


    /**
     * レイアウトのストローク幅をV5形式に変換します
     */
    private static void convertStrokeWidth(CustomControls layout) {
        for (ControlData data : layout.mControlDataList) {
            data.strokeWidth = Tools.pxToDp(computeStrokeWidth(data.strokeWidth, data.getWidth(), data.getHeight()));
        }

        for (ControlDrawerData data : layout.mDrawerDataList) {
            data.properties.strokeWidth = Tools.pxToDp(computeStrokeWidth(data.properties.strokeWidth, data.properties.getWidth(), data.properties.getHeight()));
            for (ControlData subButtonData : data.buttonProperties) {
                subButtonData.strokeWidth = Tools.pxToDp(computeStrokeWidth(subButtonData.strokeWidth, data.properties.getWidth(), data.properties.getWidth()));
            }
        }
    }

    /**
     * サイズのパーセンテージをpxサイズに変換します。古いレイアウトバージョンで使用されます
     */
    static int computeStrokeWidth(float widthInPercent, float width, float height) {
        float maxSize = Math.max(width, height);
        return (int) ((maxSize / 2) * (widthInPercent / 100));
    }
}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/ui/subassembly/customcontrols/LayoutConverter.kt
package com.arata.yukarilauncher.ui.subassembly.customcontrols

import android.content.Context
import com.google.gson.JsonSyntaxException
import com.arata.yukarilauncher.R
import com.arata.yukarilauncher.utils.LwjglGlfwKeycode
import com.arata.yukarilauncher.Tools
import org.apache.commons.io.IOUtils
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import org.lwjgl.glfw.CallbackBridge
import java.io.IOException
import java.io.InputStream
import java.nio.charset.StandardCharsets
import java.util.ArrayList

/** 旧バージョンのコントロールレイアウトJSONを最新バージョンに変換するユーティリティオブジェクト。 */
object LayoutConverter {
    /**
     * JSONファイルを読み込み、必要に応じて変換を行います。
     * @param ctx Androidコンテキスト
     * @param jsonPath JSONファイルのパス
     * @return 変換されたCustomControls、失敗時はnull
     * @throws IOException ファイル読み込みエラー
     * @throws JsonSyntaxException JSONパースエラー
     */
    @Throws(IOException::class, JsonSyntaxException::class)
    fun loadAndConvertIfNecessary(ctx: Context, jsonPath: String): CustomControls? {
        val jsonLayoutData = Tools.read(jsonPath)
        return try {
            val layoutJobj = JSONObject(jsonLayoutData)
            loadAndConvertIfNecessary(ctx, layoutJobj, jsonLayoutData, jsonPath, true)
        } catch (e: Exception) {
            Tools.showError(ctx, ctx.getString(R.string.controls_load_failed), e)
            null
        }
    }

    /**
     * アセットファイルからレイアウトを読み込みます。
     * @param context Androidコンテキスト
     * @param jsonName アセット内のJSONファイル名
     * @return 変換されたCustomControls、失敗時はnull
     */
    fun loadFromAssets(context: Context, jsonName: String): CustomControls? {
        return try {
            context.assets.open(jsonName).use { `is` ->
                val string = IOUtils.toString(`is`, StandardCharsets.UTF_8)
                val layoutJobj = JSONObject(string)
                loadAndConvertIfNecessary(context, layoutJobj, string, null, true)
            }
        } catch (e: Exception) {
            Tools.showError(context, context.getString(R.string.controls_load_failed), e)
            null
        }
    }

    /**
     * JSONObjectからレイアウトを読み込みます。
     * @param ctx Androidコンテキスト
     * @param layoutJobj レイアウトのJSONオブジェクト
     * @param jsonString 元のJSON文字列
     * @param jsonPath JSONファイルのパス（null可）
     * @param showError エラー表示を行うかどうか
     * @return 変換されたCustomControls、失敗時はnull
     */
    fun loadFromJsonObject(ctx: Context, layoutJobj: JSONObject, jsonString: String, jsonPath: String?, showError: Boolean): CustomControls? {
        return try {
            loadAndConvertIfNecessary(ctx, layoutJobj, jsonString, jsonPath, showError)
        } catch (e: Exception) {
            if (showError) Tools.showError(ctx, ctx.getString(R.string.controls_load_failed), e)
            null
        }
    }

    /**
     * 内部的な変換処理。バージョンに応じて適切な変換メソッドを呼び出します。
     * @param ctx Androidコンテキスト
     * @param layoutJobj レイアウトのJSONオブジェクト
     * @param jsonLayoutData JSON文字列
     * @param jsonPath ファイルパス
     * @param showError エラー表示フラグ
     * @return 変換されたCustomControls
     * @throws Exception 変換エラー
     */
    @Throws(Exception::class)
    private fun loadAndConvertIfNecessary(ctx: Context, layoutJobj: JSONObject, jsonLayoutData: String, jsonPath: String?, showError: Boolean): CustomControls? {
        return if (!layoutJobj.has("version")) {
            val layout = convertV1Layout(layoutJobj)
            if (jsonPath != null) layout.save(jsonPath)
            layout
        } else {
            val version = layoutJobj.getInt("version")
            when (version) {
                2 -> {
                    val layout = convertV2Layout(layoutJobj)
                    if (jsonPath != null) layout.save(jsonPath)
                    layout
                }
                3, 4, 5 -> convertV3_4Layout(layoutJobj)
                6, 7 -> convertV6_7Layout(layoutJobj)
                8 -> Tools.GLOBAL_GSON.fromJson(jsonLayoutData, CustomControls::class.java)
                else -> {
                    if (showError) {
                        val ioException = IOException()
                        Tools.showError(ctx, ctx.getString(R.string.controls_unsupported_layout_version), ioException)
                    }
                    null
                }
            }
        }
    }

    /**
     * バージョン6/7のレイアウトをバージョン8に変換します。
     * @param oldLayoutJson 旧レイアウトのJSONオブジェクト
     * @return 変換されたCustomControls
     */
    fun convertV6_7Layout(oldLayoutJson: JSONObject): CustomControls {
        val layout = Tools.GLOBAL_GSON.fromJson(oldLayoutJson.toString(), CustomControls::class.java)
        for (data in layout.mJoystickDataList!!) {
            if (data.getHeight() > data.getWidth()) {
                val ratio = data.getHeight() / data.getWidth()
                data.dynamicX = data.dynamicX!!.replace("\${height}", "($ratio * \${height})")
                data.dynamicY = data.dynamicY!!.replace("\${height}", "($ratio * \${height})") + " + (" + (ratio - 1) + " * \${height})"
                data.setHeight(data.getWidth())
            }
        }
        layout.version = 8
        return layout
    }

    /**
     * バージョン3/4のレイアウトをバージョン6に変換します。
     * @param oldLayoutJson 旧レイアウトのJSONオブジェクト
     * @return 変換されたCustomControls
     */
    private fun convertV3_4Layout(oldLayoutJson: JSONObject): CustomControls {
        val layout = Tools.GLOBAL_GSON.fromJson(oldLayoutJson.toString(), CustomControls::class.java)
        convertStrokeWidth(layout)
        layout.version = 6
        return layout
    }

    /**
     * バージョン2のレイアウトをバージョン3に変換します。
     * @param oldLayoutJson 旧レイアウトのJSONオブジェクト
     * @return 変換されたCustomControls
     * @throws JSONException JSONパースエラー
     */
    @Throws(JSONException::class)
    private fun convertV2Layout(oldLayoutJson: JSONObject): CustomControls {
        val layout = Tools.GLOBAL_GSON.fromJson(oldLayoutJson.toString(), CustomControls::class.java)
        val layoutMainArray = oldLayoutJson.getJSONArray("mControlDataList")
        layout.mControlDataList = ArrayList(layoutMainArray.length())
        for (i in 0 until layoutMainArray.length()) {
            val button = layoutMainArray.getJSONObject(i)
            val n_button = Tools.GLOBAL_GSON.fromJson(button.toString(), ControlData::class.java)
            if (!Tools.isValidString(n_button.dynamicX) && button.has("x")) {
                val buttonC = button.getDouble("x")
                val ratio = buttonC / CallbackBridge.physicalWidth
                n_button.dynamicX = "$ratio * \${screen_width}"
            }
            if (!Tools.isValidString(n_button.dynamicY) && button.has("y")) {
                val buttonC = button.getDouble("y")
                val ratio = buttonC / CallbackBridge.physicalHeight
                n_button.dynamicY = "$ratio * \${screen_height}"
            }
            layout.mControlDataList!!.add(n_button)
        }
        val layoutDrawerArray = oldLayoutJson.getJSONArray("mDrawerDataList")
        layout.mDrawerDataList = ArrayList()
        for (i in 0 until layoutDrawerArray.length()) {
            val button = layoutDrawerArray.getJSONObject(i)
            val buttonProperties = button.getJSONObject("properties")
            val n_button = Tools.GLOBAL_GSON.fromJson(button.toString(), ControlDrawerData::class.java)
            if (!Tools.isValidString(n_button.properties.dynamicX) && buttonProperties.has("x")) {
                val buttonC = buttonProperties.getDouble("x")
                val ratio = buttonC / CallbackBridge.physicalWidth
                n_button.properties.dynamicX = "$ratio * \${screen_width}"
            }
            if (!Tools.isValidString(n_button.properties.dynamicY) && buttonProperties.has("y")) {
                val buttonC = buttonProperties.getDouble("y")
                val ratio = buttonC / CallbackBridge.physicalHeight
                n_button.properties.dynamicY = "$ratio * \${screen_height}"
            }
            layout.mDrawerDataList!!.add(n_button)
        }
        convertStrokeWidth(layout)
        layout.version = 3
        return layout
    }

    /**
     * バージョン1のレガシーレイアウトをバージョン3に変換します。
     * @param oldLayoutJson 旧レイアウトのJSONオブジェクト
     * @return 変換されたCustomControls
     * @throws JSONException JSONパースエラー
     */
    @Throws(JSONException::class)
    private fun convertV1Layout(oldLayoutJson: JSONObject): CustomControls {
        val empty = CustomControls()
        val layoutMainArray = oldLayoutJson.getJSONArray("mControlDataList")
        for (i in 0 until layoutMainArray.length()) {
            val button = layoutMainArray.getJSONObject(i)
            val n_button = ControlData()
            val keycodes = intArrayOf(LwjglGlfwKeycode.GLFW_KEY_UNKNOWN.toInt(), LwjglGlfwKeycode.GLFW_KEY_UNKNOWN.toInt(), LwjglGlfwKeycode.GLFW_KEY_UNKNOWN.toInt(), LwjglGlfwKeycode.GLFW_KEY_UNKNOWN.toInt())
            n_button.dynamicX = button.getString("dynamicX")
            n_button.dynamicY = button.getString("dynamicY")
            if (!Tools.isValidString(n_button.dynamicX) && button.has("x")) {
                val buttonC = button.getDouble("x")
                val ratio = buttonC / CallbackBridge.physicalWidth
                n_button.dynamicX = "$ratio * \${screen_width}"
            }
            if (!Tools.isValidString(n_button.dynamicY) && button.has("y")) {
                val buttonC = button.getDouble("y")
                val ratio = buttonC / CallbackBridge.physicalHeight
                n_button.dynamicY = "$ratio * \${screen_height}"
            }
            n_button.name = button.getString("name")
            n_button.opacity = ((button.getInt("transparency") - 100) * -1) / 100f
            n_button.passThruEnabled = button.getBoolean("passThruEnabled")
            n_button.isToggle = button.getBoolean("isToggle")
            n_button.setHeight(button.getInt("height").toFloat())
            n_button.setWidth(button.getInt("width").toFloat())
            n_button.bgColor = 0x4d000000.toInt()
            n_button.strokeWidth = 0f
            if (button.getBoolean("isRound")) {
                n_button.cornerRadius = 35f
            }
            var next_idx = 0
            if (button.getBoolean("holdShift")) {
                keycodes[next_idx] = LwjglGlfwKeycode.GLFW_KEY_LEFT_SHIFT.toInt()
                next_idx++
            }
            if (button.getBoolean("holdCtrl")) {
                keycodes[next_idx] = LwjglGlfwKeycode.GLFW_KEY_LEFT_CONTROL.toInt()
                next_idx++
            }
            if (button.getBoolean("holdAlt")) {
                keycodes[next_idx] = LwjglGlfwKeycode.GLFW_KEY_LEFT_ALT.toInt()
                next_idx++
            }
            keycodes[next_idx] = button.getInt("keycode")
            n_button.keycodes = keycodes
            empty.mControlDataList!!.add(n_button)
        }
        empty.scaledAt = oldLayoutJson.getDouble("scaledAt").toFloat()
        empty.version = 3
        return empty
    }

    /**
     * ストローク幅の値をパーセントからdp単位に変換します。
     * @param layout 変換対象のCustomControls
     */
    private fun convertStrokeWidth(layout: CustomControls) {
        for (data in layout.mControlDataList!!) {
            data.strokeWidth = Tools.pxToDp(computeStrokeWidth(data.strokeWidth, data.getWidth(), data.getHeight()).toFloat())
        }
        for (data in layout.mDrawerDataList!!) {
            data.properties.strokeWidth = Tools.pxToDp(computeStrokeWidth(data.properties.strokeWidth, data.properties.getWidth(), data.properties.getHeight()).toFloat())
            for (subButtonData in data.buttonProperties) {
                subButtonData.strokeWidth = Tools.pxToDp(computeStrokeWidth(subButtonData.strokeWidth, data.properties.getWidth(), data.properties.getWidth()).toFloat())
            }
        }
    }

    /**
     * ストローク幅のピクセル値を計算します。
     * @param widthInPercent パーセントで指定された幅
     * @param width ボタンの幅
     * @param height ボタンの高さ
     * @return 計算されたピクセル値
     */
    fun computeStrokeWidth(widthInPercent: Float, width: Float, height: Float): Int {
        val maxSize = Math.max(width, height)
        return ((maxSize / 2) * (widthInPercent / 100)).toInt()
    }
}